src/python/pants/option/config.py

```diff
@@ -224,82 +213,46 @@ def recursively_format_str(value: str) -> str:
-                if not isinstance(member, str):
-                    return str(member)
-                interpolated_member = possibly_interpolate(member) if interpolate else member
-                return f'"{interpolated_member}"'
-
-            list_members = ", ".join(stringify_list_member(member) for member in raw_value)
-            return f"{list_prefix or ''}[{list_members}]"
-
-        return str(raw_value)
-
-    def _stringify_val_without_interpolation(self, raw_value: _TomlValue) -> str:
-        return self._stringify_val(
-            raw_value,
-            option="",
-            section="",
-            section_values={},
-            interpolate=False,
-        )
-
     def get_value(self, section: str, option: str) -> str | None:
         section_values = self.section_to_values.get(section)
         if section_values is None:
             return None
         if option not in section_values:
             return None
 
-        stringify = partial(
-            self._stringify_val,
-            option=option,
-            section=section,
-            section_values=section_values,
-        )
+        def stringify(raw_val: _TomlValue, prefix: str = "") -> str:
+            string_val = self._possibly_interpolate_value(
+                raw_value=str(raw_val),
+                option=option,
+                section=section,
+                section_values=section_values or {},
+            )
+            return f"{prefix}{string_val}"
 
         option_value = section_values[option]
         if not isinstance(option_value, dict):
             return stringify(option_value)
 
-        # Handle dict options, along with the special `my_list_option.add` and
-        # `my_list_option.remove` syntax. We only treat `add` and `remove` as the special list
-        # syntax if the values are lists to reduce the risk of incorrectly special casing.
-        has_add = isinstance(option_value.get("add"), list)
-        has_remove = isinstance(option_value.get("remove"), list)
-        if not has_add and not has_remove:
-            return stringify(option_value)
-
-        add_val = stringify(option_value["add"], list_prefix="+") if has_add else "[]"
-        remove_val = stringify(option_value["remove"], list_prefix="-") if has_remove else "[]"
-        if has_add and has_remove:
-            return f"{add_val},{remove_val}"
-        if has_add:
-            return add_val
-        return remove_val
+        # Handle dict options, along with the special `my_dict_option.add`, `my_list_option.add` and
+        # `my_list_option.remove` syntax. We only treat `add` and `remove` as the special syntax
+        # if the values have the approprate type, to reduce the risk of incorrectly special casing.
+        has_add_dict = isinstance(option_value.get("add"), dict)
+        has_add_list = isinstance(option_value.get("add"), list)
+        has_remove_list = isinstance(option_value.get("remove"), list)
+
+        if has_add_dict:
```

Comment:
Would it make sense to support `my_dict.remove`?

Reply:
Maybe, but not in this change, which already has a lot going on in it.

We'd have to figure out what the value is. A list of keys? A dict of keys to dummy values? A dict of keys to values and we only remove if the key has that value?

Reply:
https://github.com/pantsbuild/pants/issues/16482